JoinColony/block-ingestor
# Add mint tokens motion support

## Description

This pr adds block-ingestor support for the MotionCreated event, specifically when it wraps the mint tokens action.

I have coupled the MotionCreated event listeners with the Voting Reputation extension, such that as soon as it is initialised, or if `trackExtensions` determines that it is initialised on block-ingestor start up, it will also listen for MotionCreated events.

## Testing

To test, you will need to:

1. Be on this cdapp branch: https://github.com/JoinColony/colonyCDapp/pull/261. Make sure the block ingestor is commented out in in `colony-cdapp-env-orchestration.yaml`:

``` yaml

  # block-ingestor:
  #   container_name: 'ingestor'
  #   image: colony-cdapp-dev-env/block-ingestor:latest
  #   volumes:
  #     - '../amplify/mock-data:/colonyCDapp/amplify/mock-data'
  #   ports:
  #     - '10001:10001'
  #   depends_on:
  #     network-contracts:
  #       condition: service_healthy
  #     amplify:
  #       condition: service_healthy
  #   links:
  #     - "network-contracts:network-contracts.docker"
  #     - "amplify:amplify.docker"
```
2. Add reputation to the user. This can be done by:
* Copying the following into a new tab and ensuring it says the reputation miner is on: http://localhost:3001/reputation/monitor/toggle
* In a new terminal window, run `npm run truffle console`
* Paste the following:
  *  c = await IColony.at('{colonyAddress}') (where colony address is the address of the colony you're testing out)
  Then
  *  c.emitDomainReputationReward(1, accounts[0], '1000000000000000000') (make sure you're logged in with dev wallet 1).
  * Wait for the reputation mining cycle to complete and you should see the reputation update in the ui.
3. Install and initialise the Voting Reputation (Governance) extension
4. Click the "Test Mint Tokens" button. If all is set up correctly, you should be redirected to a motions screen. 

Testing should reveal that only one motion is created in the db, and it should be created in the colony you clicked "test mint tokens" in.

You should also terminate the block-ingestor and restart it, and you should see a listener attached to the MotionCreated event as a result of `trackExtensions` determining that the Voting Reputation Extn is installed.  

## Notes

In my dev env, I've been noticing that the event listener attached to the `MotionCreated` event is firing for each colony, as well as often multiple times for the same colony. 

I have guarded against the motion being associated with the wrong colony in the db, but you may see errors in the block ingestor terminal when creating a motion in the db, because it may attempt to create the same motion twice. This should not be a problem and should not actually create more than one motion per tx hash. You can verify this in the ui as well as at localhost:20002 by ensuring only one motion is created per click.
 
I have also noticed that sometimes when you click the button, it redirects but the block ingestor doesn't pick up the event and so the action page doesn't load. If this happens, go back to Colony Home and refresh & try again. We suspect this may be the result of hot reloading the block ingestor and therefore only a problem in dev. 

Finally, I've noticed that a couple of TokensMinted events get picked up straight after, with an amount of undefined and 0. I've added logic to guard against making a db mutatation when this is the case. 
  
## Troubleshooting

If you are experiencing an incorrent nonce error at colony create when running the block ingestor separately, it may be caused by the `await provider.getNetwork()` call on line 45 of index.ts. This is just a theory but commenting it out and the `getChainId()` call on the line below solved it for me the first time I tried it.
